contracts/src/Teleporter/TeleporterMessenger.sol

```diff
@@ -245,42 +232,47 @@ contract TeleporterMessenger is ITeleporterMessenger, ReentrancyGuards {
             "TeleporterMessenger: invalid origin sender address"
         );
 
-        // If the blockchain ID has yet to be initialized, do so now.
-        bytes32 blockchainID_ = blockchainID;
-        if (blockchainID_ == bytes32(0)) {
-            blockchainID_ = WARP_MESSENGER.getBlockchainID();
-            blockchainID = blockchainID_;
-        }
-
         // Parse the payload of the message.
         TeleporterMessage memory teleporterMessage =
             abi.decode(warpMessage.payload, (TeleporterMessage));
 
+        // If the blockchain ID has yet to be initialized, do so now.
+        bytes32 blockchainID_ = _initializeBlockchainID();
+
         // Require that the message was intended for this blockchain.
         require(
             teleporterMessage.destinationBlockchainID == blockchainID_,
             "TeleporterMessenger: invalid destination chain ID"
         );
 
-        // Require that the message has not been delivered previously.
-        require(
-            !_messageReceived(warpMessage.sourceChainID, teleporterMessage.messageID),
-            "TeleporterMessenger: message already delivered"
+        // Require that the message nonce is non-zero because the value is used to provide replay protection.
+        require(teleporterMessage.messageNonce != 0, "TeleporterMessenger: zero message nonce");
+
+        // Calculate the message ID of the message given the source blockchain ID and message nonce.
+        bytes32 messageID = calculateMessageID(
+            warpMessage.sourceChainID, blockchainID_, teleporterMessage.messageNonce
         );
 
+        // Require that the message has not been delivered previously.
+        require(!_messageReceived(messageID), "TeleporterMessenger: message already delivered");
+
         // Check that the caller is allowed to deliver this message.
         require(
             _checkIsAllowedRelayer(msg.sender, teleporterMessage.allowedRelayerAddresses),
             "TeleporterMessenger: unauthorized relayer"
         );
 
-        // Store the relayer reward address provided, effectively marking the message as received.
-        _relayerRewardAddresses[warpMessage.sourceChainID][teleporterMessage.messageID] =
-            relayerRewardAddress;
+        // Store the message nonce, effectively marking the message as received.
```

Comment:
think we can use a private function like `_markReceived` or `setReceived` to make it more clear

Reply:
Yeah good call. I went with `_markMessageReceived` and also moved the check that the message nonce is non-zero to be within it to make it less error prone.